### ISSUE TITLE
emergency shell on first boot w/ ignition possibly due to DMA Write faults with Dell BOSS Card

### ISSUE THREAD

**Issue by dlamotte** [coreos/bugs#2363]
# Issue Report #

## Bug ##

Failure to move ROOT to another device in ignition (following https://coreos.com/os/docs/latest/root-filesystem-placement.html).

```
  storage:
    disks:
      - device: "/dev/disk/by-path/pci-0000:18:00.0-scsi-0:2:0:0"
        partitions:
          - label: "ROOT"
        wipeTable: true

    filesystems:
      - name: ROOT
        mount:
          device: "/dev/disk/by-path/pci-0000:18:00.0-scsi-0:2:0:0"
          format: ext4
          label: ROOT

      # wipe original CoreOS root filesystem
      - name: unused
        mount:
          device: "/dev/disk/by-path/pci-0000:3b:00.0-ata-1"
          format: ext4
          label: "unused"
          wipeFilesystem: true
```

I ran `coreos-install -d /dev/disk/by-path/pci-0000:3b:00.0-ata-1 -i ignition.json` via PXE where above is a snippet of the ignition.  With the above, I'm dropped into the dracut emergency shell and `/sysroot` remains unmounted.  Worse, `ls /dev/sd*` only yields devices and not partitions (`sda sdb` vs `sda sda1 sdb sdb1 ...`).  The partition table is unable to be read with `partx -l /dev/sda` or `partx -l /dev/sdb` as well.

I see the following error in `journalctl`:

```
...
ignition[1184]: disks: createFilesystems: op(9): [finished] determining filesystem type of "/dev/disk/by-path/pci-0000:3b:00.0-ata-1"
kernel: dmar_fault: 6 callbacks suppressed
kernel: DMAR: DRHD: handling fault status reg 2
kernel: DMAR: [DMA Write] Request device [3b:00.0] fault addr ff5d7000 [fault reason 05] PTE Write access is not set
...
```

The DMAR lines repeat a bunch of times.

Finally, we get to this error:

```
...
systemd[1]: systemd-fsck-root.service: Bound to unit dev-disk-by\x2dlabel-ROOT.device, but unit isn't active.
systemd[1]: Dependency failed for File System Check on /dev/disk/by-label/ROOT.
systemd[1]: Dependency failed for /sysroot.
systemd[1]: Dependency failed for /sysroot/usr.
systemd[1]: Dependency failed for Initrd File Systems.
...
```

After this, it drops into the emergency shell.

### Container Linux Version ###
1632.2.1

### Environment ###

Bare Metal
Dell R640 w/ iDRAC9
2x Dell BOSS Card 120GB in RAID-1 (`/dev/disk/by-path/pci-0000:3b:00.0-ata-1`)
2x SAS SSD 960GB in RAID-1 via Dell PERC H740P Mini Controller (`/dev/disk/by-path/pci-0000:18:00.0-scsi-0:2:0:0`)

### Expected Behavior ###

Normal boot with `/` on SAS SSD and all other CoreOS Partitions on Dell BOSS Card.

### Actual Behavior ###

Failure to boot.  Emergency shell.

**Comment by dlamotte** [comment]
I found https://bugs.launchpad.net/ubuntu/+source/linux/+bug/911236 and it looks like its related to the `intel_iommu`.  I might try booting with `intel_iommu=off` to see if it works, but I honestly don't exactly know what that will even do.  Assuming IO performance would be slower as a result and is at most a workaround?

**Comment by dlamotte** [comment]
So I tried it with `intel_iommu=off` and the `DMA Write` errors disappeared, but Verity Setup failed for /dev/mapper/usr.  I'm booting from UEFI, so maybe I broke something else.

Is this a valid scenario?  Mounting root on a RAID-1 Mixed Use SSD and having all other CoreOS partitions sitting on Dell BOSS Read Intensive SSD (low durability)?

**Comment by bgilbert** [comment]
It is a valid scenario.  There are a couple errors in your config though:

1. In your `storage` section you're partitioning the `scsi` device, but your `filesystems` declaration creates an ext4 filesystem over the entire disk.
2. Your `unused`  filesystem is overwriting the entire boot disk, not just the root filesystem.

So, in both cases, be sure to specify the correct partition and not the entire disk.  Modifying your example to use labels (which is a bit cleaner):

```
  storage:
    disks:
      - device: "/dev/disk/by-path/pci-0000:18:00.0-scsi-0:2:0:0"
        partitions:
          # use partition label distinct from existing ROOT partition
          - label: "NEW-ROOT"
        wipeTable: true

    filesystems:
      # wipe original CoreOS root filesystem
      - name: unused
        mount:
          device: "/dev/disk/by-partlabel/ROOT"
          format: ext4
          label: "unused"
          wipeFilesystem: true

      - name: ROOT
        mount:
          device: "/dev/disk/by-partlabel/NEW-ROOT"
          format: ext4
          # filesystem label (must be ROOT)
          label: ROOT
```

**Comment by dlamotte** [comment]
Perfect.  This makes sense to me.  I'll give it a try and update this issue.  Might be a few days before I get to it.

**Comment by dlamotte** [comment]
So, I tried this and `ignition-disks` fails during first boot but there is no clear error.  I can see the proper path (`/dev/disk/by-path/pci-0000:18:00.0-scsi-0:2:0:0`) in the emergency shell.  But for some reason, `ignition-disks` gives up with exit status 1.

Failing that, I attempted to do the equivalent as the ignition steps in a shell script during PXE (before ignition's first boot attempt) in hopes that it would work.  It didn't.

Unfortunately, I think I have to give up on this at this point.  But I made some notes and would definitely like to tackle this in the future at some point and/or if someone else manages to figure it out.

Thanks for the pointers!
